### DISCUSSION TITLE
Status of GDL on Apple Silicon ?

### DISCUSSION THREAD

**Discussion by alaingdl** [gnudatalanguage/gdl#1187] [General]
Do we have any report of success of GDL compilation on Apple with the new M1 processors ?

I did received one negative feedback, without details

**Comment by brandy125** [comment]
Tested it right now and compiled/run normally:

***@***.***:~/gdln/gdl-master/build/src$ uname -a
Linux ubuntu 5.8.0-56-generic #63~20.04.1-Ubuntu SMP Tue Jun 8 01:47:26 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

***@***.***:~/gdln/gdl-master/build/src$ ./gdl
  GDL - GNU Data Language, Version 1.0.1 Git
- For basic information type HELP,/INFO
- Default library routine search path used (GDL_PATH/IDL_PATH env. var. not set): /usr/local/share/gnudatalanguage/lib
- Using WxWidgets as graphics library (windows and widgets).
- No startup file read (GDL_STARTUP/IDL_STARTUP env. var. not set). 
- Please report bugs, feature or help requests and patches at: https://github.com/gnudatalanguage/gdl

GDL> print,!GDL
{ 1.0.1 Git Oct 21 2021  1634785200   1   1   1}
GDL>

**Comment by Jeanne-jw** [comment]
I met the problem like this:
for way 0: brew install gnudatalanguage: No available formula or cask with the name "gnudatalanguage".
for way 1: In my Mac ( MacOS Big Sur 11.6, Apple M1) the path of CMAKE_C_COMPILER and CMAKE_CXX_COMPILER are separately "/opt/homebrew/opt/llvm/bin/clang" and "/opt/homebrew/opt/llvm/bin/clang++", they are different with what they are assumed to be in build_gdl.sh, which are "/usr/local/opt/llvm/bin/clang" and "/usr/local/opt/llvm/bin/clang++".
So I just changed the PATH in the file build_gdl.sh and then it worked.  
Will this cause problems in later procession ?

**Comment by alaingdl** [comment]
Way 0 : ha, sorry, gnudatalanguage formula is no more available in Brew, we have to correct that in the wiki.

Way 1 : did you already go a message from the script _build_gdl.sh_ ? I received positive feedback recently for OSX 11.x (not sure of the sub-version) and in the past, at some point, one is a link on the other one ...

**Comment by slayoo** [comment]
https://github.com/brewsci/homebrew-science/blob/master/Formula/gnudatalanguage.rb

**Comment by slayoo** [comment]
It seems to be still there:
https://github.com/brewsci/homebrew-science/blob/master/Formula/gnudatalanguage.rb

but first you need to ad the science tap:
```
$ brew tap brewsci/science 
```
or
```
$ brew tap homebrew/science
```

**Comment by Jeanne-jw** [comment]
Thank you very much!
Yes, after taping
$ brew tap brewsci/science
the formula gnudatalanguage can be found!

**Comment by GillesDuvert** [comment]
yes but it is 0.9.7, in eh, 2017? and we are 1.0.1

**Comment by alaingdl** [comment]
Concerning the compilation of current version, no problem on OSX with 1.0.1 git version with qhull (`brew install qhull`) on OSX 11.6 and the script build_gdl.sh

**Comment by Jeanne-jw** [comment]
I realize that using Homebrew for M1, packages are installed into the /opt/homebrew/ folder. But x86_64 Homebrew is still usable on M1 Mac, and using this version of homebrew, the path would be /usr/local/homebrew, which would match the path in build_gdl.sh 
Referencing website: https://medium.com/mkdir-awesome/how-to-install-x86-64-homebrew-packages-on-apple-m1-macbook-54ba295230f#bf20

**Comment by Barney-9000** [comment]
Just downloaded/compiled today's GDL 1.0.1 using the three stage `bash scripts/build_gdl.sh` method on both macOS 10.14.6 (iMac Intel) and macOS 12.0.1 (M1 Pro MBP). Compilations appeared to progress identically, with /usr/local/opt/llvm/bin at the front of the $path in both cases. The macOS 12.0.1 GDL is much (much) slower for certain operations.  

TIME_TEST3 / TIME_TEST4
0.351882=Total Time / 3.98409=Total Time > Mojave 10.14.6
6.58272=Total Time / 78.4071=Total Time > Monterey 12.0.1

Full listing below with the standout performance hits for Monterey picked out. Loops seem to be the killer.  It was a READ_ASCII operation which first alerted me to compare the two.

**macOS 10.14.6**

GDL> TIME_TEST3
% Compiled module: TIME_TEST3.
% Compiled module: TIME_TEST.
|TIME_TEST3 performance for IDL 8.2:
|       OS_FAMILY=unix, OS=darwin, ARCH=x86_64
| Fri Nov 12 16:21:09 2021
       1    0.0366070 Empty For loop, 2000000 times
       2    0.0107110 Call empty procedure (1 param) 100000 times
       3    0.0368030 Add 200000 integer scalars and store
       4    0.0403459 50000 scalar loops each of 5 ops, 2 =, 1 if)
       5  0.000715017 Mult 512 by 512 byte by constant and store, 30 times
       6   0.00489998 Shift 512 by 512 byte and store, 300 times
       7   0.00148511 Add constant to 512x512 byte array, 100 times
       8   0.00156403 Add two 512 by 512 byte arrays and store, 80 times
       9   0.00212502 Mult 512 by 512 floating by constant, 30 times
      10   0.00397611 Shift 512 x 512 array, 60 times
      11   0.00322485 Add two 512 by 512 floating images, 40 times
      12    0.0133801 Generate 1000000 random numbers
      13   0.00409102 Invert a 192^2 random matrix
      14   0.00606918 LU Decomposition of a 192^2 random matrix
      15    0.0228150 Transpose 384^2 byte, FOR loops
      16    0.0207369 Transpose 384^2 byte, row and column ops x 10
      17    0.0169189 Transpose 384^2 byte, TRANSPOSE function x 100
      18    0.0244021 Log of 100000 numbers, FOR loop
      19   0.00149298 Log of 100000 numbers, vector ops 10 times
      20    0.0177310 131072 point forward plus inverse FFT
      21    0.0124960 Smooth 512 by 512 byte array, 5x5 boxcar, 10 times
      22   0.00993299 Smooth 512 by 512 floating array, 5x5 boxcar, 5 times
      23    0.0593588 Write and read 512 by 512 byte array x 40
     0.351882=Total Time,     0.0083620777=Geometric mean,      23 tests.
GDL> TIME_TEST4
% Compiled module: TIME_TEST4.
|TIME_TEST4 performance for IDL 8.2:
|       OS_FAMILY=unix, OS=darwin, ARCH=x86_64
| Fri Nov 12 16:21:38 2021
       1    0.0723901 Empty for loop, 6000000 times
       2     0.407959 Foreach, 6000000 elements
       3    0.0311620 Call empty procedure (1 param) 300000 times
       4     0.109629 Add 600000 integer scalars and store
       5     0.115855 150000 scalar loops each of 5 ops, 2 =, 1 if)
       6   0.00200105 Mult 512 by 512 byte by constant and store, 90 times
       7    0.0142920 Shift 512 by 512 byte and store, 900 times
       8   0.00440884 Add constant to 512x512 byte array, 300 times
       9   0.00452399 Add two 512 by 512 byte arrays and store, 240 times
      10   0.00692606 Mult 512 by 512 floating by constant, 90 times
      11    0.0109189 Shift 512 x 512 array, 180 times
      12   0.00904298 Add two 512 by 512 floating images, 120 times
      13    0.0140619 Generate 3000000 random numbers
      14    0.0188439 Invert a 332^2 random matrix
      15    0.0123630 LU Decomposition of a 332^2 random matrix
      16    0.0972240 Transpose 665^2 byte, FOR loops
      17    0.0464098 Transpose 665^2 byte, row and column ops x 10
      18    0.0450919 Transpose 665^2 byte, TRANSPOSE function x 100
      19    0.0714722 Log of 300000 numbers, FOR loop
      20   0.00392580 Log of 300000 numbers, vector ops 10 times
      21     0.172625 2097152 point forward plus inverse FFT
      22    0.0360539 Smooth 512x512 byte array, 5x5 boxcar, 30 times
      23    0.0557289 Smooth 512x512 floating array, 5x5 boxcar, 30 times
      24      2.34015 Write and read 512x512 byte array, 120 times
      25     0.281038 Create 120000 empty lists
      3.98409=Total Time,      0.034006231=Geometric mean,      25 tests.

-----------------------------------------------

**macOS 12.0.1**

GDL> time_test3
% Compiled module: TIME_TEST3.
% Compiled module: TIME_TEST.
|TIME_TEST3 performance for IDL 8.2:
|       OS_FAMILY=unix, OS=darwin, ARCH=x86_64
| Fri Nov 12 16:20:52 2021
       1    0.0202389 Empty For loop, 2000000 times
       2   0.00474596 Call empty procedure (1 param) 100000 times
       **3 >>>     1.78803 Add 200000 integer scalars and store**
       **4  >>>     2.24613 50000 scalar loops each of 5 ops, 2 =, 1 if)**
       5  0.000885963 Mult 512 by 512 byte by constant and store, 30 times
       6    0.0173490 Shift 512 by 512 byte and store, 300 times
       7   0.00184202 Add constant to 512x512 byte array, 100 times
       8   0.00171089 Add two 512 by 512 byte arrays and store, 80 times
       9   0.00183606 Mult 512 by 512 floating by constant, 30 times
      10   0.00535798 Shift 512 x 512 array, 60 times
      11   0.00223804 Add two 512 by 512 floating images, 40 times
      12    0.0128679 Generate 1000000 random numbers
      13   0.00232410 Invert a 192^2 random matrix
      14   0.00397110 LU Decomposition of a 192^2 random matrix
      **15   >>>    1.32873 Transpose 384^2 byte, FOR loops**
      16    0.0821719 Transpose 384^2 byte, row and column ops x 10
      17    0.0377369 Transpose 384^2 byte, TRANSPOSE function x 100
      18     0.902741 Log of 100000 numbers, FOR loop
      19   0.00237608 Log of 100000 numbers, vector ops 10 times
      20    0.0124781 131072 point forward plus inverse FFT
      21    0.0137341 Smooth 512 by 512 byte array, 5x5 boxcar, 10 times
      22    0.0211489 Smooth 512 by 512 floating array, 5x5 boxcar, 5 times
      23    0.0720739 Write and read 512 by 512 byte array x 40
      6.58272=Total Time,      0.018329246=Geometric mean,      23 tests.
GDL> time_test4
% Compiled module: TIME_TEST4.
|TIME_TEST4 performance for IDL 8.2:
|       OS_FAMILY=unix, OS=darwin, ARCH=x86_64
| Fri Nov 12 16:21:23 2021
       1    0.0614450 Empty for loop, 6000000 times
       **2 >>>      53.4269 Foreach, 6000000 elements**
       3    0.0136681 Call empty procedure (1 param) 300000 times
       **4 >>>      5.40138 Add 600000 integer scalars and store**
       **5 >>>     6.70570 150000 scalar loops each of 5 ops, 2 =, 1 if)**
       6   0.00204086 Mult 512 by 512 byte by constant and store, 90 times
       7    0.0529270 Shift 512 by 512 byte and store, 900 times
       8   0.00487804 Add constant to 512x512 byte array, 300 times
       9   0.00475192 Add two 512 by 512 byte arrays and store, 240 times
      10   0.00515795 Mult 512 by 512 floating by constant, 90 times
      11    0.0156569 Shift 512 x 512 array, 180 times
      12   0.00668120 Add two 512 by 512 floating images, 120 times
      13    0.0182812 Generate 3000000 random numbers
      14    0.0107679 Invert a 332^2 random matrix
      15   0.00570679 LU Decomposition of a 332^2 random matrix
      **16 >>>     3.98595 Transpose 665^2 byte, FOR loops**
      17     0.154704 Transpose 665^2 byte, row and column ops x 10
      18    0.0836458 Transpose 665^2 byte, TRANSPOSE function x 100
      19      2.68452 Log of 300000 numbers, FOR loop
      20   0.00826693 Log of 300000 numbers, vector ops 10 times
      21     0.132512 2097152 point forward plus inverse FFT
      22    0.0405221 Smooth 512x512 byte array, 5x5 boxcar, 30 times
      23     0.123635 Smooth 512x512 floating array, 5x5 boxcar, 30 times
      24      2.99870 Write and read 512x512 byte array, 120 times
      **25  >>>     2.45869 Create 120000 empty lists**
      78.4071=Total Time,      0.092388868=Geometric mean,      25 tests.

**Comment by GillesDuvert** [comment]
@Barney-9000 , this is very interesting and probably points to some drawbacks of the current implementation of GDL.
However the for ... do loop in GDL per se is not so slow. The first test, "empty for loop"  is quite comparable.
For the record, on my machine I have:
```
GDL> .compile time_test
GDL> tic & for i=0,6000000 do time_test_dummy & toc
% Time elapsed :       0.38093114 seconds.
```
```
IDL> .compile time_test      
IDL> tic & for i=0,6000000 do time_test_dummy & toc
% Time elapsed: 0.45998096 seconds.
```
with the unexpected result that GDL is even marginally faster for 'bare' loops than IDL...
It is when something is done inside the loop that things go awry, but not always. This is interesting because 'doing things' requires the creation/destruction of variables, it is there that GDL is disfavoured. 

However, if you used build_gdl.sh, is not it a chance that some or all of the libraries you retrieved (from BREW) were not compiled for M1? Meaning they will be 'emulated' by the machine, and then its speed will go from average to worse?

**Comment by alaingdl** [comment]
((problems to post with my old Firefox))

I tested on a Mac today (troubles concerning the version, seems to be in 12.0.1 ?!)

For me the problable reason  is the usage of OpenMP, or not, 
With current OSX (and LLVM), if is quite difficult to activate OpenMP

On several multicores Linux I have, I don't have such discrepancies between IDL and GDL on time_test3/4,
times are in line ...

**Reply by Barney-9000** [reply]
Thanks for the tips. Indeed, all the brew stuff was migrated over from an Intel Mac. I did a full update, but now realise that M1 brew is a little different.  So, I'll do a full M1 install of Homebrew ([helpful guide](https://mac.install.guide/homebrew/)), recompile a fresh version of GDL and let you know. Probably won't get round to it until next week.

**Comment by Barney-9000** [comment]
Had a bit of time. OK, now twice as fast - but still very slow on the `Foreach, 6000000 elements` loop. I tweaked the `build_gdl.sh` script to pickup the new homebrew location `/opt/homebrew`, but must have missed something:

`PLplot installation lacks the requested driver: wxwidgets`

So no plotting just yet.  Here's the `time_test4` listing:

```
GDL> time_test4

% Compiled module: TIME_TEST4.
% Compiled module: TIME_TEST.
|TIME_TEST4 performance for IDL 8.2:
|       OS_FAMILY=unix, OS=darwin, ARCH=arm64
| Wed Nov 17 18:25:22 2021
       1    0.0459521 Empty for loop, 6000000 times
       2      22.5409 Foreach, 6000000 elements
       3    0.0104151 Call empty procedure (1 param) 300000 times
       4      2.29468 Add 600000 integer scalars and store
       5      2.86319 150000 scalar loops each of 5 ops, 2 =, 1 if)
       6   0.00103903 Mult 512 by 512 byte by constant and store, 90 times
       7    0.0243351 Shift 512 by 512 byte and store, 900 times
       8   0.00288105 Add constant to 512x512 byte array, 300 times
       9   0.00315189 Add two 512 by 512 byte arrays and store, 240 times
      10   0.00340199 Mult 512 by 512 floating by constant, 90 times
      11   0.00887012 Shift 512 x 512 array, 180 times
      12   0.00418091 Add two 512 by 512 floating images, 120 times
      13    0.0173771 Generate 3000000 random numbers
      14    0.0100849 Invert a 332^2 random matrix
      15   0.00778198 LU Decomposition of a 332^2 random matrix
      16      1.70346 Transpose 665^2 byte, FOR loops
      17    0.0763040 Transpose 665^2 byte, row and column ops x 10
      18    0.0351219 Transpose 665^2 byte, TRANSPOSE function x 100
      19      1.14478 Log of 300000 numbers, FOR loop
      20   0.00266004 Log of 300000 numbers, vector ops 10 times
      21    0.0991251 2097152 point forward plus inverse FFT
      22    0.0412810 Smooth 512x512 byte array, 5x5 boxcar, 30 times
      23    0.0556588 Smooth 512x512 floating array, 5x5 boxcar, 30 times
% Compiled module: FILEPATH.
      24     0.105150 Write and read 512x512 byte array, 120 times
      25      1.36851 Create 120000 empty lists
      32.4703=Total Time,      0.048331505=Geometric mean,      25 tests.
```

**Comment by Barney-9000** [comment]
Attempted various ways to get PLplot to work -- no joy, I'm afraid.  Anyone managed to get this working for a native M1 install?

**Comment by alaingdl** [comment]
Did you look on plplot pages ? https://sourceforge.net/p/plplot/support-requests/

**Comment by Barney-9000** [comment]
Nothing that I can see.  Brew installs PLplot and wxwidgets without complaint.  Nevertheless, PLplot (invoked from GDL) can't seem to find the wxwidgets driver.  Could it be that the pre-Monterey homebrew paths `/usr/local` are hardwired somewhere and need to be updated with `/opt/homebrew`?

BTW, at the final stage of the build this warning repeats many times:
```
ld: warning: direct access in function 'lib::file_search(EnvT*)' from file 'CMakeFiles/gdl.dir/file.cpp.o' to global weak symbol 'typeinfo for Data_<SpDString>' from file 'CMakeFiles/gdl.dir/datatypes.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```

**Reply by GillesDuvert** [reply]
not finding the plplot driver is the classic nightmare. perhaps it not even present in the homebrew distribution, to start with?

**Comment by Barney-9000** [comment]
Any way to check?  The file plplot/drivers.h is present in both the Mojave and Monterey homebrew installs, and the only reference to wxwidgets is the same: `PLDLLIMPEXP_DRIVER void plD_dispatch_init_wxwidgets( PLDispatchTable *pdt );`

**Reply by GillesDuvert** [reply]
what the plplot engine needs, when invoked by GDL to create a wxWidgets plot window, is a small shared lib named 'wxwidgets.so', on my machine it is in /usr/local/lib/plplot5.11.1/drivers/wxwidgets.so (/local/ because I compiled plplot myself).
This file comes with plplot distributions (ubuntu, debian etc) but not necessary automatically, and more than often in a separate package (e.g., plplot-driver-wxwidgets). Often it is needed to install it separately, 
This may be true for homebrew.
Additionnally, GDL's plplot may not be aware of where the driver has been installed, the 'knowledge' probably being lost in thecreation of the plplot package. In that case, it is possible to set the env. var. PLPLOT_DRV_DIR to where the wxwidgets.so file resides.
The PLPLOT documentation is not clear about this I think.

**Reply by Barney-9000** [reply]
Thanks for the tips. But now I really am stumped: I can't find `wxwidgets.so` anywhere on the Mojave GDL 1.0.1 install, which works fine. That was installed using the three stage  `scripts/build_gdl.sh` method, and the graphics are definitely wx (see below). But there is no sign of either the `drivers` directory or  `wxwidgets.so` on both the Mojave and Monterey installs... 

```
GDL> gdl_status
% Compiled module: GDL_STATUS.
% Compiled module: GDL_IDL_FL.
 1 DSFMT ?  :        1
 2 Eigen3 ? :        1
 3 Expat ?  :        1
 4 FFTw ?   :        1
 5 GEOTIFF ?:        1
 6 GLPK ?   :        1
 7 GRIB ?   :        1
 8 HDF ?    :        0
 9 HDF5 ?   :        1
10 Magick ? :        1
11 NetCDF ? :        1
12 NetCDF4 ?:        1
13 OpenMP ? :        1
14 PNGLIB ? :        1
15 PROJ ?   :        1
16 Python ? :        1
17 ShapeLib :        1
18 TIFF ?   :        1
19 UDUNITS  :        1
20 WxWidgets:        1
21 X11      :        0
Some _EXISTS() functions are missing !
  Expected      :  22
  Really tested :  21
GDL> print,!gdl
{ 1.0.1 Git Nov 12 2021  1636675200   1   1   1}
```

**Reply by GillesDuvert** [reply]
On OSX build_gdl.sh recompiles plplot with wxWidgets, see [this](https://github.com/gnudatalanguage/gdl/blob/master/scripts/deps/macos/brew_enable_wxwidgets)
May be there was an issue there in the new architecture, as the formula may have changed?

**Comment by alaingdl** [comment]
I assume the current way for OSX in CI is for Intel ? Is it difficult to prepare one for M1 ? 

Another point is we can try to compile on OSX with `Clang --target=arm64-apple-darwin` even on a Intel proc. ...
(no time to test up to now :slightly_frowning_face: (and my usual test bed on OSX is now broken since last update of OSX)

**Comment by Barney-9000** [comment]
I wonder if it is Monterey rather than M1: with macOS 12 all homebrew resides in /opt/homebrew.

**Comment by alaingdl** [comment]
macOS 12 : since the update I have : 

```xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun```

and I cannot go ahead (a remote machine).
[

**Comment by Barney-9000** [comment]
Maybe [this](https://mac.install.guide/homebrew/3.html) explains the xcrun problem? See section: **After a macOS Upgrade**.

**Comment by GillesDuvert** [comment]
@btom54  I was intrigued by 1). Can you give the contents of test.pro ? I fail to reproduce it.
2) can at the moment be solved only by compiling with gcc or perhaps follow [this](https://github.com/BLLIP/bllip-parser/issues/19)
3) certainly hits yet another OO deficiency of GDL, each time we correct one of them more OO procedures work OK, so it would be nice to have a working example of... this failure. Such as the gridded_profiles__define.pro for example, or a simpler version.

**Comment by GillesDuvert** [comment]
Hi,
I've run your example with idl and gdl using the following files:
[test.txt](https://github.com/gnudatalanguage/gdl/files/7785510/test.txt)
[test.pro.txt](https://github.com/gnudatalanguage/gdl/files/7785513/test.pro.txt)
(.pro files cannot be attached in GH, I had to rename it).

The thing is, if I do not add `test` after `.run test.pro`, I get the same result with IDL and GDL, namely:
```
% Compiled module: TEST.
% Attempt to call undefined procedure: 'CELLCH4'.
% Execution halted at: $MAIN$          
% Attempt to call undefined procedure: 'CH4'.
% Execution halted at: $MAIN$          
% Attempt to call undefined procedure: 'CO2'.
% Execution halted at: $MAIN$          
% Attempt to call undefined procedure: 'H2O'.
% Execution halted at: $MAIN$          
       3950.0000
       4550.0000
       1
      24
```
which is GDL or IDL trying to print a something using 'implied print'.

If now I add `test` and so really call the test.pro procedure, then the results are the same:
```
$ idl < test.txt 
(...)
% Compiled module: TEST.
cellCH4
CH4
CO2
H2O
       3950.0000       4550.0000
       1
      24
% Stop encountered: TEST               29 /tmp/test.pro
```
```
$ gdl < test.txt 
% Compiled module: TEST.
cellCH4
CH4
CO2
H2O
       3950.0000       4550.0000
       1
      24
% Stop encountered: TEST                29 /tmp/test.pro
```
However I managed to reproduce, sort of, the `% End of file encountered. Unit: 0` with this test.pro code:
```
pro test
  x=''
  while not eof(0) do begin
     read, x
     print,'read:',x
  endwhile
end
```
meaning that at minimum there is a problem with eof() on unit 0.

**Comment by GillesDuvert** [comment]
actually EOF(0) now works after a 3 lines correction in basic_fun.cpp

**Comment by Barney-9000** [comment]
Anyone managed to progress with compiling GDL on M1 / Monterey?  I'm happy to try out any suggestions / testing.  Note, I'm a user of GDL/IDL rather than a developer, and have never coded in C.  So, can't help with the nitty-gritty, I'm afraid.

**Reply by alaingdl** [reply]
With @Jeanne-jw we did succeed. We try hard to have all the options, and to use the script, but no success. We did compiled using _classical_ cmake after adding as much as possible packages using Brew (through the script). We also had a trouble with readline, not well detected (TBC)

I am trying to find our up-to-date notes, it seems that it was OK with :

```
cmake .. -DHDF=off -DOPENMP=off -DPYTHON=off -DX11=on -DPLPLOTDIR=/Users/jw/GDL0103/PLplot/plplot-5.15.0/Compilation -DREADLINEDIR=/opt/homebrew/Cellar/readline/8.1.2/
```

Oh, yes, we did compiled by our-self plplot, but is was easy.

**Comment by Barney-9000** [comment]
Thanks for the tip.  I just needed to add `-DWXWIDGETS=off` and then it worked like a charm (with just a smattering of complaints about 'weak symbol ...' at the final linking stage). Much appreciated!

**Comment by alaingdl** [comment]
Great ! In fact I was in touch with @Jeanne-jw this afternoon trying to confirm whether `-DWXWIDGETS=off` was mandatory or not.

**Comment by pjb7687** [comment]
My student @djk0706 just successfully compiled GDL on M1 machine without disabling wxwidgets, we will make a PR soon.

**Reply by Barney-9000** [reply]
Excellent!  Tried it myself over the weekend, with no success -- so looking forward to hear how it can be done.  Did you manage to compile with OpenMP enabled?

**Reply by pjb7687** [reply]
As far as I remember we didn't disable anything else but let me check it again tomorrow.

**Reply by im-going-nowng** [reply]
![image](https://user-images.githubusercontent.com/101169016/182626979-fae70944-ec71-46dd-a704-3f366c07da3f.png)
Yes. I compiled with openMP enabled.

**Reply by pjb7687** [reply]
I've just merged #1352, please ping me if anything is wrong. Thanks!

**Reply by Barney-9000** [reply]
Ah, perhaps should have replied here:

Compiled the source from [djk0706:master](https://github.com/djk0706/gdl) yesterday on my M1 (macOS 12.5) - went smoothly with just a few warnings.  Ran a few of my routines with no problems, though did seem slower than the same version (1.0.1) on an Intel iMac (macOS 11.6.8). The `TIME_TEST4` procedure on both revealed this was indeed the case:
` 1.66323=Total Time` Intel iMac (macOS 11.6.8)
`32.9407=Total Time` M1 MBP (macOS 12.5) 
See also #1187.   As before, it's the same tests that run slow:

`       2      22.9813 Foreach, 6000000 elements`
`       4      2.30151 Add 600000 integer scalars and store`
`      5      2.89165 150000 scalar loops each of 5 ops, 2 =, 1 if)`
`      16      1.72188 Transpose 665^2 byte, FOR loops`
`     19      1.15830 Log of 300000 numbers, FOR loop`
`      25      1.35935 Create 120000 empty lists`

Do you also see this?

**Reply by pjb7687** [reply]
@Barney-9000 Could you quickly check the file size of your gdl binary? Just to make sure it is compiled in "Release" mode.

**Reply by Barney-9000** [reply]
16,231,536 bytes

**Reply by anaguirre** [reply]
If anyone has gotten GDL to compile on an up-to-date (Monterey M1) I'd be incredibly grateful if you could post the exact set of commands you used to do so?  (Or if you have any idea what's going wrong.) I've been trying for a while now; it dies at the very end of the compile with:

ld: warning: ignoring file /Library/Frameworks/proj.framework/proj, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
Undefined symbols for architecture arm64:
  "_pj_errno", referenced from:
      lib::map_init(DStructGDL*) in projections.cpp.o
  "_pj_fwd", referenced from:
      lib::gdlProjForward(void*, DStructGDL*, Data_<SpDDouble>*, Data_<SpDDouble>*, Data_<SpDLong>*, bool, Data_<SpDLong>*&, bool, Data_<SpDLong>*&, bool, bool) in projections.cpp.o
      lib::gdlApplyFullProjection(void*, DStructGDL*, Data_<SpDDouble>*, Data_<SpDDouble>*) in projections.cpp.o
      lib::protect_proj_fwd(lib::projUV, void*) in projections.cpp.o
      lib::gdlFullProjectionTransformation(void*, DStructGDL*, Data_<SpDDouble>*, Data_<SpDDouble>*) in projections.cpp.o
  "_pj_init", referenced from:
      lib::map_init(DStructGDL*) in projections.cpp.o
  "_pj_init_plus", referenced from:
      lib::map_init(DStructGDL*) in projections.cpp.o
  "_pj_inv", referenced from:
      lib::map_proj_inverse_fun(EnvT*) in projections.cpp.o
      lib::protect_proj_inv(lib::projUV, void*) in projections.cpp.o
  "_pj_strerrno", referenced from:
      lib::map_init(DStructGDL*) in projections.cpp.o
ld: symbol(s) not found for architecture arm64
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/gdl] Error 1
make[1]: *** [src/CMakeFiles/gdl.dir/all] Error 2
make: *** [all] Error 2

**Reply by alaingdl** [reply]
Should be OK following #1352

My way in the past was to start using the build script (to have as much as possible dependances)
then to move to cmake and desactivate some tricky parts (X11, wxwidgets ...) then go ahead ...

We did succeed with my student in winter on M1 ... with X11 but no WXwidgets

_bon courage_

A.

**Comment by GillesDuvert** [comment]
See #1498 . As of jan 2023 GDL has changed a lot and most of the comments here are now moot.
But on M1 it is still not completely ok.
